File: medium/python/2638.py
// https://leetcode.com/problems/count-the-number-of-k-free-subsets/description/

class Solution:
    def countTheNumOfKFreeSubsets(self, nums: List[int], k: int) -> int:
        nums.sort()

        tail = Counter()
        for num in nums:
            if num - k in tail:
                cnt = tail[num - k]
                del tail[num - k]
                tail[num] = cnt + 1
            else:
                tail[num] = 1

        n = max(tail.values()) + 2
        dp = [0] * (n)
        dp[0] = dp[1] = 1
        for i in range(2, n):
            dp[i] = dp[i - 1] + dp[i - 2]

        res = 1
        for v in tail.values():
            res *= dp[v + 1]

        return res        

File: medium/cpp/2638.cpp
// https://leetcode.com/problems/count-the-number-of-k-free-subsets/description/

class Solution
{
public:
    long long countTheNumOfKFreeSubsets(std::vector<int>& nums, int k)
    {
        std::unordered_map<int, int> tail;

        std::sort(std::begin(nums), std::end(nums));
        auto n = 1;

        for (auto num : nums)
        {
            if (auto it = tail.find(num - k); it != std::end(tail))
            {
                const auto old = it->second;
                tail.erase(it);
                tail.emplace(num, old + 1);
                n = std::max(n, old + 1);
            }
            else
            {
                tail.emplace(num, 1);
            }
        }

        std::vector<long long> dp(n + 2);
        dp[0] = 1;
        dp[1] = 1;

        for (auto i = 2; i <= n + 1; i++)
        {
            dp[i] = dp[i - 1] + dp[i - 2];
        }

        auto res = 1LL;

        for (const auto& [val, cnt] : tail)
        {
            res *= dp[cnt + 1];
        }

        return res;
    }
};